src/build/urlfetcher.go

```diff
@@ -0,0 +1,277 @@
+	cacheDir string
+	client   *http.Client
+	noCache  bool
+}
+
+// URLInfo is a metadata representing stored or to-be-stored url
+type URLInfo struct {
+	ID       string
+	URL      string
+	FileName string `json:"-"`
+	BaseName string
+	HasEtag  bool
+	Etag     string
+	Size     int64
+	Fetcher  *URLFetcherFS `json:"-"`
+}
+
+// NewURLFetcherFS returns an instance of URLFetcherFS, initialized to
+// live in <base>/url_fetcher_cache
+func NewURLFetcherFS(base string, noCache bool, httpClient *http.Client) (cache *URLFetcherFS) {
+	cacheDir := filepath.Join(base, "url_fetcher_cache")
+
+	return &URLFetcherFS{
+		cacheDir: cacheDir,
+		client:   httpClient,
+		noCache:  noCache,
+	}
+}
+
+// GetInfo retrieves stored URLInfo data
+func (uf *URLFetcherFS) GetInfo(url0 string) (info *URLInfo, err error) {
+	info, ok, err := uf.getURLInfo(url0)
+	if err != nil {
+		return nil, err
+	}
+
+	if !ok {
+		return nil, fmt.Errorf("no url found in cache: `%s`", url0)
+	}
+
+	return info, nil
+}
+
+// Get downloads url, stores file and metadata in cache
+func (uf *URLFetcherFS) Get(url0 string) (info *URLInfo, err error) {
+	info, ok, err := uf.getURLInfo(url0)
+	if err != nil {
+		return nil, err
+	}
+
+	if !uf.noCache && ok {
+
+		log.Debugf("Validating %s [%s]", info.URL, info.FileName)
+
+		if info.isEtagValid() {
+			log.Debugf("%s valid!", info.URL)
+
+			return info, nil
+		}
+	}
+
+	err = info.download()
+	if err != nil {
+		return nil, err
+	}
+
+	return info, nil
+}
+
+func (uf *URLFetcherFS) getURLInfo(url0 string) (info *URLInfo, ok bool, err error) {
+	info, err = uf.makeURLInfo(url0)
+	if err != nil {
+		return nil, false, err
+	}
+
+	ok, err = info.load()
+	if err != nil {
+		return nil, false, err
+	}
+
+	return info, ok, nil
+}
+
+func (uf *URLFetcherFS) makeURLInfo(u string) (info *URLInfo, err error) {
+	if !isURL(u) {
+		return nil, fmt.Errorf("expecting http:// or https:// url, got `%s` instead", u)
+	}
+
+	u1, err := url.Parse(u)
+	if err != nil {
+		return nil, err
+	}
+
+	baseName := filepath.Base(u1.Path)
+
+	if baseName == "" {
+		return nil, fmt.Errorf("unable to determine filename from url: %s", u)
+	}
+
+	h := sha256.Sum256([]byte(u))
+	id := fmt.Sprintf("%x", h)
+
+	info = &URLInfo{
+		ID:       id,
+		URL:      u,
+		BaseName: baseName,
+		Fetcher:  uf,
+	}
+	info.FileName = info.getBlobFileName()
+	return info, nil
+}
+
+func (uf *URLFetcherFS) makeID(u string) (id string) {
+	h := sha256.Sum256([]byte(u))
+	id = fmt.Sprintf("%x", h)
+	return id
+}
+
+func isURL(u string) bool {
+	return (7 <= len(u) && u[:7] == "http://") ||
+		(8 <= len(u) && u[:8] == "https://")
+}
+
+func (info *URLInfo) getBlobFileName() (fileName string) {
+	return filepath.Join(info.Fetcher.cacheDir, info.ID[:2], info.ID)
+}
+
+func (info *URLInfo) getInfoFileName() (fileName string) {
+	return info.getBlobFileName() + ".json"
+}
+
+func (info *URLInfo) isEtagValid() bool {
+	if !info.HasEtag {
+		return false
+	}
+
+	var httpClient *http.Client
```

Comment:
You can move the fallback to the default http client to the `NewURLFetcherFS` constructor.